### PR TITLE
Fix worker action deserialization race condition

### PR DESCRIPTION
**Pull request by ghale**
This fixes a race condition where errors when deserializing the worker action could show up during either `WorkerProcess.start()` or `WorkerProcess.waitForStop()`.  They will now always show up during `start()`.

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcess.java [hunk 1]
```diff
@@ -165,11 +173,12 @@ private void doStart() {
                     throw UncheckedException.throwAsUncheckedException(e);
                 }
             }
-            if (processFailure != null) {
-                throw UncheckedException.throwAsUncheckedException(processFailure);
-            }
```
Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcess.java]: Why is this moved into the `connection == null` case? I probably do not grasp the complete lifecycle. Are we certain that we will not miss any `processFailure` now?

Comment (ghale) [inline subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcess.java]: Yeah, it's really about determining when the failure has occurred and making sure the error is surfaced during the correct method.  We're really trying to determine if the failure occurs before "startup" has finished, or after.  If we get to this point and the connection has not been made yet, then the failure occurred before we finished startup and we throw the exception during `start()`.  On the other hand, if the connection has been made, then it occurred after startup and we want to throw the exception during `waitForStop()` (even though we already know that it failed).  This goes back to adam's point that failures of a particular type should happen reliably in either `start()` or `waitForStop()`, but not both.

Comment (jjohannes) [inline subprojects/core/src/main/java/org/gradle/process/internal/worker/DefaultWorkerProcess.java]: Ok. Now I got it. Thanks.
